Links that open in a new tab/window need to inform users a new tab/window will open

mehigh opened the pull request:
## Description
Fixes #1105 by adding the external dashicon and the standard screen reader text too.

## How Has This Been Tested?
I checked that the Excerpt external link which makes use of the new ExternalLink component contains the new icon and displays it correctly in major OSX browsers (Safari, Chrome, Firefox). 
I used the browser's inspector to confirm the screen-reader-text element gets properly added as well.

## Screenshots:
![Updated panel](http://files.urldocs.com/share/excerpt-block-updated/excerpt-block-updated.png)

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation. (Other as basic components don't have inline documentation, just like this one)

iseulde commented:
Is this the correct way to do this? I wish we could use CSS and use `target="_blank"`, otherwise, could we make a shared component for this?

afercia commented:
> could we make a shared component for this?

In another project at Yoast we've actually built a simple `<NewTabMessage>` component. It's a simple span with a translatable text and visually hidden by default. No icon in our case. Here, the icon could be passed as a prop, maybe.

mehigh commented:
@iseulde 
The icon can come through via CSS, but the added mark-up (the screen reader extra text) is required either way. And since this is the only occurrence in the Guttenberg interface, I think it can live fine by itself as it is until there's a need for a second usage.

afercia commented:
I think `PostPermalink` uses a second `target="_blank"`.

mehigh commented:
Thank you, afercia. I'll see whether adding the external icon looks OK in the PostPermalink area, and if it does I'll add it in and also create a component for them.

mehigh commented:
> I think PostPermalink uses a second target="_blank".

I looked all over the interface of Guttenberg, but I failed to find the PostPermalink area.

I checked how it's done in the regular posts editing screen:
![Permalink](http://files.urldocs.com/share/permalink/permalink.png)
And the page loads in the same tab, hence given the standard functionality an external icon is not required on that element.

mehigh commented:
I've exported the newWindowIcon into its own component, but I didn't find a way to use it the 2nd time yet.

iseulde commented:
I'm sure there will be. What I actually meant is something like `<ExternalLink href="..." />`.

mehigh commented:
I've stopped using the inline SVG and instead used CSS to add the icon and it offers an improved UX because now the icon too changes color on hover to match the color of the link text.

Icon matches link text color:
![Icon](http://files.urldocs.com/share/Screen-Shot-2017-07-01-at-16.21.22/Screen-Shot-2017-07-01-at-16.21.22.png)

iseulde commented:
I think we do want to keep the SVG instead of using the icon font? Cc @jasmussen

mehigh commented:
Change of heart @iseulde ? (you previously mentioned "I wish we could use CSS"). 
With an SVG file inlined, there's no reason to use any CSS whatsoever then for the icon.

I saw the inlined SVG icons are used almost everywhere, but also saw that the dashicons font is enqueued in the guttenberg page, so... let me know and I'll add the SVG back in or not when I go about creating the ExternalLink component to replace the externalIcon I just created.

iseulde commented:
Yeah, I _wish_ we could use CSS. 🙂

jasmussen commented:
Let's avoid the icon font if we can. Pretty sure this ticket is a duplicate, and Andrew Duthie, had an idea for how to add "external" as a property that would simply output the icon after the link. It sounded pretty good to me, and so long as the js code is simple, it's cool.

jasmussen commented:
Doh, sorry, didn't realize this was a PR referencing the very issue I was talking about. My apologies, I blame being mobile.

mehigh commented:
Understood, jasmussen. Yes, the iconfont can be avoided. I'll update the PR accordingly.

mehigh commented:
I've resurrected the usage of the SVG and dropped the iconfont according to @jasmussen 's input.
I also created the <ExternalLink href="https://example.com">LinkText</ExternalLink> component according to @iseulde 's input. This covers the addition of the icon and the accessibility text.

Thank you for all your recommendations.

iseulde commented on components/external-link/index.js:
```diff
@@ -0,0 +1,24 @@
+	return (
+		<a { ...additionalProps } href={ href } target="_blank">
+			{ children }
+			<span className="new-window-icon">
```
Is there a need for this `span`? Can the class not be put on the parent?

mehigh replied:
@iseulde 
This isn't needed, I'll update the PR to include it on the anchor tag instead and drop the span.

iseulde commented on components/external-link/style.scss:
```diff
@@ -0,0 +1,7 @@
+		height: 1.4em;
+		vertical-align: top;
+	}
+}
```
We don't have checks for it, but: https://github.com/WordPress/gutenberg/blob/v0.3.0/.editorconfig#L9. Is there any way to add checks @nylen or @ntwb?

ntwb replied:
Yes, it's checked as part of the default stylelint config:
> https://github.com/WordPress-Coding-Standards/stylelint-config-wordpress/blob/master/index.js#L71

mehigh commented:
I've handled the latest update and also updated the PR description to match the PULL_REQUEST_TEMPLATE.md I've seen, such that it's easier to get all of this in.

youknowriad commented on components/external-link/index.js:
```diff
@@ -0,0 +1,28 @@
+import Dashicon from '../dashicon';
+import './style.scss';
+
+function ExternalLink( { href, children, className, ...additionalProps } ) {
```
Should we drop `href` (it's not used here, and could be included by `...additionalProps`

mehigh replied:
I intentionally passed that through the href={ href } because without it the ExternalLink becomes useless (opens up a new empty tab). I know it's not a _real_ required parameter, but it felt better this way.

nylen commented:
I'd say leave `href`, it's such a fundamental property of a link that you
want to see it in the code.

On Jul 3, 2017 11:05 AM, "Riad Benguella" <notifications@github.com> wrote:

> *@youknowriad* commented on this pull request.
> ------------------------------
>
> In components/external-link/index.js
> <https://github.com/WordPress/gutenberg/pull/1577#discussion_r125240982>:
>
> > + * External dependencies
> + */
> +import classnames from 'classnames';
> +
> +/**
> + * WordPress dependencies
> + */
> +import { __ } from 'i18n';
> +
> +/**
> + * Internal dependencies
> + */
> +import Dashicon from '../dashicon';
> +import './style.scss';
> +
> +function ExternalLink( { href, children, className, ...additionalProps } ) {
>
> Should we drop href (it's not used here, and could be included by
> ...additionalProps
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/1577#pullrequestreview-47610829>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAN2zsJsUEu54Jh0SNyuwZsj6dhwD8OVks5sKK7MgaJpZM4OI85u>
> .
>

mehigh commented:
@iseulde any steps I can take to allow this development to more forward and not get stale?

mehigh commented:
I've refreshed the PR with the latest in master (a small conflict was popping up).

aduth commented on components/external-link/style.scss:
```diff
@@ -0,0 +1,7 @@
+.new-window-icon {
+	.dashicon {
```
A bit of horizontal margin might be nice, but is minor and can be addressed in a styling pass if necessary.

mehigh replied:
Thanks. I'll make that happen. Good point with the noopener/etc. rels.

aduth reviewed (changes requested):
This is looking quite good, but I'd suggest the addition of a `rel` attribute.

First for the `external` link type, see also:

https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types

Second for `noopener` and `noreferrer`, see also:

- https://dev.to/ben/the-targetblank-vulnerability-by-example
- https://github.com/Automattic/wp-calypso/pull/7680

We can can join with incoming prop value:

```js
function ExternalLink( { href, children, className, rel = '', ...additionalProps } ) {
	rel = uniq( compact( [
		...rel.split( ' ' ),
		'external',
		'noreferrer',
		'noopener'
	] ) ).join( ' ' );
```

https://lodash.com/docs/4.17.4#uniq
https://lodash.com/docs/4.17.4#compact

mehigh committed:
Links that open in a new tab/window need to inform users a new tab/window will open

## Description
Fixes #1105 by adding the external dashicon and the standard screen
reader text too.

## How Has This Been Tested?
I checked that the Excerpt external link which makes use of the new
ExternalLink component contains the new icon and displays it correctly
in major OSX browsers (Safari, Chrome, Firefox).
I used the browser's inspector to confirm the screen-reader-text
element gets properly added as well.

## Screenshots:
![Updated
panel](http://files.urldocs.com/share/excerpt-block-updated/excerpt-bloc
k-updated.png)

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation. (Other as basic
components don't have inline documentation, just like this one)

mehigh commented:
@aduth 

Your input has been handled. I've also squashed my changes in a single commit and updated this to the latest in master.
I've added the rels and a small margin on the sides of the icon - looks better this way.

Thanks for the resource links, they've been really handy.

mehigh committed:
1105 - Fix formatting

aduth commented on components/external-link/index.js:
```diff
@@ -0,0 +1,35 @@
+		'noreferrer',
+		'noopener',
+	] ) ).join( ' ' );
+	const classes = classnames( 'new-window-icon', className );
```
Per recommended CSS Naming guidelines, the default class name should be `components-external-link`

https://github.com/WordPress/gutenberg/blob/master/docs/coding-guidelines.md#naming

aduth commented on components/external-link/index.js:
```diff
@@ -0,0 +1,35 @@
+	return (
+		<a { ...additionalProps } className={ classes } href={ href } target="_blank" rel={ rel }>
+			{ children }
+			<span className="screen-reader-text">{ __( '(opens in a new window)' ) }</span>
```
Minor: Elsewhere in the WordPress codebase, this string is preceded with a translator comment `translators: accessibility text`. My main concern would be if omitting the translator comment would create a separate string entry in GlotPress (I'm not sure)

https://translate.wordpress.org/projects/wp/dev/fr/default?filters%5Bterm%5D=opens+in+a+new+window&filters%5Buser_login%5D=&filters%5Bstatus%5D=current_or_waiting_or_fuzzy_or_untranslated&filter=Filter&sort%5Bby%5D=priority&sort%5Bhow%5D=desc

Otherwise this could be:

```jsx
<span className="screen-reader-text">
	{
		/* translators: accessibility text */
		__( '(opens in a new window)' )
	}
</span>
```

aduth reviewed (commented):
This is looking good, and I think could be merged after the class name change.

mehigh committed:
1105 - Changed the class name. Added translators: accessibility text comment

mehigh commented:
@aduth 
Feedback handled.

aduth reviewed (approved):
Looks great, thanks!
